domain-knowledge/DOMAIN-OBJECTS.md

```diff
@@ -0,0 +1,501 @@
+audio file, which is required, the service converts it to a FLAC file for long-term storage. Once
+stored in FLAC format, the audio file can be converted to any other format, such
+as Microsoft WAVE (WAV), MPEG-3 (MP3), Apple Lossless (ALAC), Advanced Audio Codec (AAC), etc.
+
+The service favors the FLAC format for a number of reasons:
+
+- FLAC is unencumbered by the many patents and royalty fees that plague competing formats.
+- FLAC encoding and decoding tools are cost-free and open-source software.
+- FLAC is technically superior to most, if not all, competing formats, offering
+ the most features and flexibility.
+ 
+When customers purchase music from the catalog, they are able to download their
+purchased music in FLAC format, in addition to several other formats:
+
+- WAV
+- MP3
+- AAC
+
+#### Relationships
+
+- `FlacFile` `hasMany()` `Song`
+
+#### Access Policy
+
+- Any `User` that is associated with a `CatalogEntity` whose `approved_at` field
+  is non-null, and whose `deleted_at` field is null, may create a `FlacFile`.
+- Anybody may read a `FlacFile` record from the _database_, but not necessarily
+  the corresponding binary file from the filesystem.
+- Only customers who have purchased access to a given `FlacFile` may retrieve
+  the associated binary data from the filesystem.
+- The `User` who is associated with a given `CatalogEntity` may delete a `FlacFile`
+  that is under its control, but only under the following conditions:
+  - The `FlacFile` is being overwritten and is not associated with any other `Song`.
+    - If the `FlacFile` _is_ associated with another `Song`, the `User` is
+      prompted: "The audio file that you're replacing is associated with other
+      songs. Which of those songs, if any, should be updated with this audio file?"
+      (The `User` is presented with a list, and appropriate UI controls, to
+      select which `Songs'` relationships, if any, should be updated.) If the
+      `User` selects every `Song`, the "orphaned" `FlacFile` is deleted, along
+      with its binary filesystem data.
+  - A `Song` is deleted that would cause the `FlacFile` to be "orphaned"; in such
+    cases, the `FlacFile` is deleted, along with the associated binary filesystem
+    data.
+
+### Genre
+
+One or more `Genres` may be assigned to an `Album`. Assigning one or
+more `Genres` to an `Album` enables the `Album` to appear in search results for
+the associated `Genre`, thereby making it easier for potential customers to find
+music in their preferred `Genre`.
+
+The service provides a pre-set list of `Genres` derived from the ID3v1 MP3 tag
+specification. Additionally, `Artists` may request that new `Genres` are added,
+which, upon request, are queued for administrative approval. This policy prevents
+`Artists` from adding duplicate `Genres` (i.e., with alternate spellings) and
+from using gratuitous profanity, inconsistent formatting, etc.
+
+#### Relationships
+
+- `Genre` `belongsToMany()` `Album`
+
```

Comment:
Can artists labels have genres too?

Reply:
That's a great question/point, and I don't see why not! I'll make that change.